### ISSUE TITLE
Red Screens on Nvidia GPUs

### ISSUE THREAD

**Issue by hyperion51** [prime31/Nez#44]
I've been using Nez. Love it. But I ran my build on my laptop and it came up all red, like GB were missing and R is all that was left. Same on each of the 3 other machines I could find at the time - all laptops with nvidia cards, all coming up red. Then I tried building the samples on my laptop and the same thing happened even there.

Searching for those symptoms reveals https://github.com/mono/MonoGame/issues/1495 this really old issue which shouldn't really apply to DesktopGL but I can't help but shake the feeling that some issue in the precompiled shaders provided with Nez is causing this, considering that the last fix on that page involves tweaking a shader. Silly question: have you actually tested Nez on nvidia?

**Comment by prime31** [comment]
Nez uses the default MonoGame [sprite shader](https://github.com/prime31/nez/blob/master/Nez-PCL/Graphics/Batcher/Batcher.cs#L72). An issue like what you are seeing where green and blue channels just arent getting rendered at all sounds like a MonoGame bug. The default MonoGame sprite shader and the shaders included with Nez would never only output the red channel.

You should be able to confirm this by just sticking a plain old SpriteBatch in there and see if it has the same result with the same textures. Also make sure you are not setting the `color` of any sprites to red which would end up tinting everything red.

**Comment by Freakmiko** [comment]
I can confirm this and I have an nvidia card as well. Here is the code that makes this happen for me.

```
public override void initialize()
        {
            base.initialize();

            clearColor = Color.Black;

            var grid = createEntity("grid");
            var springGrid = grid.addComponent(new SpringGrid(new Rectangle(-50, -50, 2020, 1180), new Vector2(30)));
            springGrid.gridMajorColor = new Color(30, 30, 139, 85);
            springGrid.gridMinorColor = new Color(30, 30, 139, 85);

            addPostProcessor(new BloomPostProcessor(3)).settings = BloomSettings.presetSettings[0];
        }
```

Without the PostProcessor it looks like this [http://i.imgur.com/wZAKql8.png](http://i.imgur.com/wZAKql8.png) and with the PostProcessor enabled it looks like this [http://i.imgur.com/04J2QXx.png](http://i.imgur.com/04J2QXx.png). This is on Windows 10 with an nvidia 770 gtx and the default effects.

**Comment by prime31** [comment]
So, is it only with bloom that you are seeing the issue? If so, that narrows it down greatly.

**Comment by Freakmiko** [comment]
Unfortunately it also happens when I'm using the Vignette instead of the Bloom.

**Comment by witlock** [comment]
Same happens with me.

**Comment by prime31** [comment]
Super odd. Does it occur outside of Nez as well (for example just stick one of the shaders in a regular MonoGame project)? I'm wondering if this is a render target bug or something else altogether.

**Comment by prime31** [comment]
Super weird. The only thing I can think of at this point is something to do with the RenderTarget swapping Nez does. I whipped up a little demo project to see if we can narrow this down. It works as both a Nez and a MonoGame app. To toggle between the two just open Game1.cs and comment/uncomment out the first line.

When in Nez mode:
- pressing 1 will disable post processing
- pressing 2 will enable post processing

When in MonoGame mode:
- pressing 1, 2, 3, 4 will cycle through the bloom modes

[KLJKLJKJL.zip](https://github.com/prime31/Nez/files/364943/KLJKLJKJL.zip)

**Comment by Freakmiko** [comment]
Yes, it seems really weird. Using the project you gave, I get a red screen in Nez whether the post processing is disabled or enabled ([Disabled](http://i.imgur.com/nVpDXgY.png), [Enabled](http://i.imgur.com/xPyXBgM.png)). The Picture in MonoGame mode is just as you'd expect ([Monogame](http://i.imgur.com/pmjSfN8.png)).

I also found out that if the post processing is disabled in the `initialize()` function like this
```
public override void initialize()
{
    createEntity( "ent" ).transform.setPosition( 200, 200 ).entity
        .addComponent( new Sprite( content.Load<Texture2D>( "sponge" ) ) );

    addPostProcessor(new BloomPostProcessor(0));
    enablePostProcessing = false;
}
```
the screen will stay normal when the game starts and when the post processing is enabled in-game with 2, the bloom effect works ([working](http://i.imgur.com/tdt0Fo5.png)).

**Comment by prime31** [comment]
Thanks for testing this stuff out. I'm down to only ATI cards at the moment.

This is really, really, really odd now. There shouldnt be any reason at all the turning post processing off in initialize and back on again later should change the way things work. Any needed RenderTargets are created when addPostProcessor is called regardless if enablePostProcessing is true or not!

A couple random things to try (total shots in the dark here):
- call `Screen.preferredDepthStencilFormat = DepthFormat.None;` at startup before the Scene is created.
- in Game.Initialize call `Window.AllowUserResizing = true;` then just try resizing the window and see if that changes anything (all RenderTargets are recreated when the window size changes)
- try using the attached Nez.dll (just copy it over the current one and then Clean your project). It has one change: RenderTargets are changed from PreserveContents to DiscardContents.

[Nez.dll.zip](https://github.com/prime31/Nez/files/365066/Nez.dll.zip)

**Comment by Freakmiko** [comment]
Sadly none of these fixed the problem (neither each individually nor all together).

It also seems to work if the first scene that's loaded doesn't have a post processor added and the next scene that's loaded does. For example, I have a menu scene without any post processors and then when I switch into the game scene, the post processors work without a hitch. 

On the other hand, if the first scene has a post processor added, all the scenes that get loaded afterwards that have a post processor (including the first scene of course) have the red screen bug.

**Comment by prime31** [comment]
Alright, so that is at least a hacky workaround: load up a fake "loading" Scene then your real Scene after that. Better than nothing I suppose. Something really fishy is going on with the RenderTarget creation and swapping here. 

I made one more little tester. It's a thermonuclear shader that doesn't do anything other than setting the red channel to 0. You can add it with a PostProcessor like this:

`addPostProcessor( new PostProcessor( 0, content.Load<Effect>( "NoMoreRed.mgfxo" ) ) );`

If that still outputs red I'm going to drive to nVidia headquarters and drop a flaming bag of poop on their front door.

[NoMoreRed.mgfxo.zip](https://github.com/prime31/Nez/files/365128/NoMoreRed.mgfxo.zip)

**Comment by prime31** [comment]
OK, I just had a really silly idea that might just work. Try adding the following snippet *before* the initial Scene is set. This will prime a couple of RenderTargets and cycle through then discard them. Perhaps that will make the gods at nVidia happy?

```
			var rt1 = RenderTarget.create();
			var rt2 = RenderTarget.create();
			graphicsDevice.setRenderTarget( rt1 );
			graphicsDevice.setRenderTarget( rt2 );
			graphicsDevice.setRenderTarget( rt1 );
			graphicsDevice.setRenderTarget( null );
			rt1.Dispose();
			rt2.Dispose();
```

**Comment by Freakmiko** [comment]
Using the `NoMoreRed` shader, I just get a black screen when the postProcessing is enabled in `initialize()`. Everything else that applied to the other post processors, applies to this one as well (like, that scenes loaded afterwards still are rendered red only).
 The second solution, cycling through RenderTargets, does not seem to work (I am supposed to call it right before the first `scene = Scene.createWithDefaultRenderer`, am I correct?).

**Comment by prime31** [comment]
So, the render texture or final blur to the backbuffer seem to only have one color channel. The shader removed the red channel a it ended up having no color at all (black). I have no idea how or why single channel textures are getting made or why an empty Scene before a real Scene fixes it or why it only happens with post processors. Nothing makes any sense!

**Comment by prime31** [comment]
Many thanks for your assistance and at least finding a workaround. I am super curious what the heck is happening here. I'm going to see if I can get a hold of a computer with an nVidia card that has the issue so that I can examine it in more detail in some graphical debugging tools.

**Comment by Scellow** [comment]
I don't have this issue at all, i'm using a GTX 960 with driver version: 368.39

Samples are all fine

**Comment by Shadoh702** [comment]
Hey guys I'm having this redness problem when assigning shaders, everything becomes red. NVidia GTX970 driver version 368.39. I'm using the latest MonoGame, but your provided example does the same:
![image](https://cloud.githubusercontent.com/assets/7041004/17759315/5b72f89c-654a-11e6-9965-2dc55b0458bf.png)
![image](https://cloud.githubusercontent.com/assets/7041004/17759319/605576b4-654a-11e6-8881-cb51932be559.png)

**Comment by prime31** [comment]
I still only have access to 2 nVidia GPUs at the moment and can't seem to repro this one. If you can figure out what triggers the red craziness (specific shader, state, etc) that would be a big help figuring out what the issue might be.

**Comment by Shadoh702** [comment]
Update: it doesn't do it if I take out the #define nez. meaning that it's possibly the way Nez does the shader apply? But it happens in my MonoGame project which applies it during a sprite batch. I'll keep looking in to it. P.S I updated my gpu driver to the latest, but no change.

**Comment by Shadoh702** [comment]
Sorry guys, I had to work-around it. My effect was a fullscreen effect, so instead of using a spritebatch to render to the effect to the screen, I used a Quad (four vertices). That solved the problem. Sorry I couldn't find a proper fix.

**Comment by Avataren** [comment]
I'm having the same exact issue ( GTX 770, 365.19 , latest nez and monogame compiled from source). I noticed something peculiar when testing that might be of interest. 
It's probably a similar "fix" as the dummy scene mentioned earlier in the thread. 

I made a button toggle the Scene.enablePostProcessing at runtime. If the scene has enablePostProcessing == true set from start, then the output will remain all red even when I toggle the postprocessing off. The rendertarget has been permanently broken after this it seems.

If I set enablePostProcessing == false from the start, and then enable it with the toggle at runtime, everything works as expected. 

It appears that some subsystem has to be initialized before the shaders will work without breaking the textures, but I haven't figured out exactly what it is.

**Comment by Avataren** [comment]
If I start my Game class (derived from Core) like this, the problem goes away..

        protected override void Initialize()
        {
            base.Initialize();
            scene = Scene.createWithDefaultRenderer();
            base.Update(new GameTime());
            base.BeginDraw();
            base.Draw(new GameTime());
            base.EndDraw();

           //if I set my real scene here, it all works
}

**Comment by prime31** [comment]
That is some good detective work! I can't say I have any idea why that works but it's at least a start in finding out why.

**Comment by Avataren** [comment]
This issue has now disappeared on my setup. 
I don't know if this is due to changes in the framework, or if it's the latest driver that fixed it (372.90), but the red only textures are gone. I have not made any changes to my hardware.

**Comment by prime31** [comment]
That's fantastic. Can anyone else who had the red texture problems confirm?

**Comment by Avataren** [comment]
I'm sorry, but I spoke too soon on this issue. Turns out I had pushed a version of the scene without postprocessing from another machine earlier, and pulled it without realizing it was off. (It's quite subtle). Issue still persists.

**Comment by Noxalus** [comment]
Hello everyone 😄 

I don't use Nez, but I have the exact same issue (red "filter") on an Android emulator when I use a post process effect (bloom). I didn't figure out what happen yet, but it seems to be related to MonoGame itself and not Nez 😟

**Comment by piedoom** [comment]
This is a stretch, but how are you using monogame?  Did you use the windows installer, or are you using NuGet?  I found issues with effects when linking to my local monogame install, but the problem was resolved when I used NuGet monogame instead.

**Comment by Noxalus** [comment]
I used the Windows installer. I just tried to use the Nugget Package instead, but it doesn't seem to fix the issue for me :/

Here is what I've got:

![monogame-pp-issue](https://cloud.githubusercontent.com/assets/1321964/20034525/77e76458-a3c1-11e6-9f83-cc20250b4203.jpg)

I've succeeded to have a proper rendering initializing a specific class with different values, but I don't understand why, it shouldn't have any impact of the post processing :s I'm still trying to understand what happen...

**Comment by coryfinnegan** [comment]
I can confirm that using base.Update(new GameTime()); base.Draw(new GameTime()); resolved the red screen issue.

**Comment by zerodarkzone** [comment]
This still happens on Android, and the proposed fix doesn't work any idea how to solve this on Android?
thanks

**Comment by xposure** [comment]
I don't know if this helps but I experienced this issue years ago with MonoGame and it still seems to exist. I once went through the series of creating the twin stick shooter game that uses bloom heavily and once I added that effect, the scene would randomly be red.

Its been so long, but I *think* what fixed it was making sure to actually render something to the target first which might be why the `base.Draw()` methods fix it? I know its not a lot of help, but I just remember that if I didn't do something in a specific order with the RTs they would simply be corrupt and locked to red no matter what you do.
